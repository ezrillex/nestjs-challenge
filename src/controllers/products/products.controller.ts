import { Controller, Get } from '@nestjs/common';
import { PublicPrivate } from '../../decorators/public_and_private/public_and_private.decorator';

@Controller('products')
export class ProductsController {
  @Get()
  @PublicPrivate()
  async getProducts() {
    return 'hello1!!!';
  }
}
