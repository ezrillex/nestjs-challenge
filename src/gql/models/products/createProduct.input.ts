import {
  ArrayMinSize,
  IsArray,
  IsString,
  ValidateNested,
} from 'class-validator';
import { Type } from 'class-transformer';
import { CreateVariationInput } from './createVariation.input';
import { Field, InputType } from '@nestjs/graphql';

@InputType()
export class CreateProductInput {
  @IsString()
  @Field()
  name: string;

  @IsString()
  @Field()
  description: string;

  @Field(() => [CreateVariationInput])
  @IsArray()
  @ValidateNested({ each: true })
  @ArrayMinSize(1)
  @Type(() => CreateVariationInput)
  variations: CreateVariationInput[];
}
