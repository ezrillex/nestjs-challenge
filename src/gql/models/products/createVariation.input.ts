import { IsArray, IsInt, IsNumber, IsString, IsUUID } from 'class-validator';
import { Field, InputType } from '@nestjs/graphql';
import { Images } from './images/images';

@InputType()
export class CreateVariationInput {
  @IsString()
  @Field()
  title: string;

  @Field()
  @IsNumber()
  price: number;

  @Field()
  @IsInt()
  stock: number;

  @Field(() => [String], { nullable: true })
  // @IsArray()
  // @IsUUID('all', { each: true })
  images: string[];
}
