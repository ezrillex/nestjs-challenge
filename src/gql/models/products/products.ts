import { Field, ObjectType } from '@nestjs/graphql';
import { IsUUID } from 'class-validator';
import { ProductVariations } from './product-variations/product-variations';

@ObjectType()
export class Products {
  @Field(() => String)
  id: string;

  @Field(() => Boolean)
  is_published: boolean;

  @Field(() => Boolean)
  is_deleted: boolean;

  @Field(() => String)
  name: string;

  @Field(() => String)
  description: string;

  @Field(() => [ProductVariations])
  variations: ProductVariations[];

  @Field(() => String)
  created_by: string;

  @Field(() => Date)
  created_at: string;

  @IsUUID()
  @Field(() => String, { nullable: true })
  last_updated_by: string;

  @Field(() => String, { nullable: true })
  last_updated_at: string;
}
