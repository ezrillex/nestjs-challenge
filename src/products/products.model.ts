import { Field, ID, ObjectType } from '@nestjs/graphql';
import { IsUUID } from 'class-validator';
import { Categories } from '../categories/categories.model';
import { ProductVariations } from './product_variation/product-variations';
import { LikesOfProducts } from '../likes/likes_of_products';

@ObjectType()
export class Products {
  @Field(() => ID)
  id: string;

  @Field(() => Boolean)
  is_published: boolean;

  @Field(() => Boolean)
  is_deleted: boolean;

  @Field(() => String)
  name: string;

  @Field(() => String)
  description: string;

  @Field(() => [LikesOfProducts])
  likes: LikesOfProducts[];

  @Field(() => [ProductVariations])
  variations: ProductVariations[];

  @Field(() => [Categories])
  categories: Categories[];

  @Field(() => String)
  created_by: string;

  @Field(() => Date)
  created_at: string;

  @IsUUID()
  @Field(() => String, { nullable: true })
  last_updated_by: string;

  @Field(() => String, { nullable: true })
  last_updated_at: string;
}
