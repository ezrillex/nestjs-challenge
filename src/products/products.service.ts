import {
  BadRequestException,
  Injectable,
  InternalServerErrorException,
  NotFoundException,
} from '@nestjs/common';
import { PrismaService } from '../prisma/prisma.service';
import { roles } from '@prisma/client';
import { CreateProductInput } from './inputs/createProduct.input';
import { GetProductsInput } from './inputs/get-products.input';
import { UpdateProductInput } from './inputs/update-product.input';
import { UpdateProductVariationInput } from './product_variation/update-product-variation.input';
import { CreateProductVariationInput } from './product_variation/create_product_variation.input';

@Injectable()
export class ProductsService {
  constructor(private prisma: PrismaService) {}

  async CreateProduct(data: CreateProductInput, userId: string) {
    return this.prisma.products.create({
      data: {
        name: data.name,
        description: data.description,
        categories: {
          connect: data.categories.map((id) => ({ id: id })),
        },
        variations: {
          createMany: {
            data: data.variations,
          },
        },
        created_by: userId,
        created_at: new Date().toISOString(),
      },
    });
  }

  async UpdateProduct(data: UpdateProductInput, userId: string) {
    // todo you could in theory update a deleted product if you have the uuid...
    const countProducts = await this.prisma.products.count({
      where: { id: data.id },
    });
    if (countProducts === 0) {
      throw new NotFoundException('Product not found.');
    }

    // assume user id is correct due to guard. no checks.

    if (
      !(data.name || data.description || data.categories || data.is_published)
    ) {
      throw new BadRequestException(
        'At least one property is required for update!',
      );
    }

    const toUpdate = {};
    if (data.name) {
      toUpdate['name'] = data.name;
    }
    if (data.description) {
      toUpdate['description'] = data.description;
    }
    if (data.categories) {
      toUpdate['categories'] = {
        set: data.categories.map((id) => ({ id: id })),
      };
    }
    if (data.is_published) {
      toUpdate['is_published'] = data.is_published;
    }

    return this.prisma.products.update({
      where: { id: data.id },
      data: {
        ...toUpdate,
        last_updated_by: userId,
        last_updated_at: new Date().toISOString(),
      },
    });
  }

  async GetProducts(role: roles, params: GetProductsInput) {
    const filter = {};
    const pagination = { skip: 0, take: 10 };

    if (params.first) {
      pagination.take = params.first;
    }

    if (params.offset) {
      pagination.skip = params.offset;
    }

    if (params.categoryFilter && params.categoryFilter.length > 0) {
      filter['categories'] = {
        some: {
          id: {
            in: params.categoryFilter,
          },
        },
      };
    }

    if (role === roles.manager) {
      filter['is_deleted'] = false;
    } else {
      // public role
      filter['is_published'] = true;
      filter['is_deleted'] = false;
    }

    if (params.likedOnly && role === roles.customer) {
      // todo filter by liked for customers.
    }

    if (params.search && params.search.length > 0) {
      filter['name'] = { contains: params.search };
    }
    // todo RBAC of queryable fields.
    return this.prisma.products.findMany({
      include: {
        variations: {
          include: { images: true },
        },
        categories: true,
      },
      where: filter,
      ...pagination,
    });
  }

  async GetProductById(role: roles, id: string) {
    const filter = { id: id };

    if (role === roles.manager) {
      filter['is_deleted'] = false;
    } else {
      // public role
      filter['is_published'] = true;
      filter['is_deleted'] = false;
    }

    return this.prisma.products.findUnique({
      include: {
        variations: {
          include: { images: true },
        },
        categories: true,
      },
      where: filter,
    });
  }

  async GetProductVariationById(id: string, count_only: boolean = false) {
    if (count_only) {
      return this.prisma.productVariations.count({
        where: { id: id },
      });
    } else {
      return this.prisma.productVariations.findUnique({
        include: {
          images: true,
        },
        where: { id: id },
      });
    }
  }

  async UpdateProductVariation(
    data: UpdateProductVariationInput,
    userId: string,
  ) {
    // todo you could in theory update a deleted product variation if you have the uuid...
    if (!(data.title || data.stock || data.price)) {
      throw new BadRequestException('Must have at least one field to update!');
    }

    const toUpdate = {};
    if (data.title) {
      toUpdate['title'] = data.title;
    }
    if (data.price) {
      toUpdate['price'] = data.price;
    }
    if (data.stock) {
      toUpdate['stock'] = data.stock;
    }

    return this.prisma.productVariations.update({
      where: { id: data.id },
      data: {
        ...toUpdate,
        last_updated_by: userId,
        last_updated_at: new Date().toISOString(),
      },
    });
  }

  async CreateProductVariation(
    data: CreateProductVariationInput,
    userId: string,
  ) {
    // todo check if product exists, or let it fail?
    return this.prisma.productVariations.create({
      data: {
        product: { connect: { id: data.product_id } },
        price: data.price,
        stock: data.stock,
        title: data.title,
        last_updated_by: userId,
        last_updated_at: new Date().toISOString(),
      },
    });
  }

  async DeleteProductVariation(id: string) {
    const record = await this.prisma.productVariations.findUnique({
      where: { id: id },
      include: {
        product: true,
      },
    });

    if (!record) {
      throw new NotFoundException('Product Variation not found.');
    }

    // todo research how to get the count directly from query.
    const variations_of_product = await this.prisma.productVariations.findMany({
      where: { product_id: record.product_id },
    });

    if (variations_of_product.length === 1) {
      throw new BadRequestException(
        'Can not delete the product variation if it is the last one.',
      );
    }

    const result = await this.prisma.productVariations.delete({
      where: { id: id },
    });

    if (result) {
      return 'Product Variation deleted.';
    } else
      throw new InternalServerErrorException(
        'Something went wrong when deleting the product variation.',
      );
  }

  async DeleteProduct(product_id: string, user_id: string) {
    // TODO because the db will cascade when we do the actual delete, we need to first delete the images from CDN.

    return this.prisma.products.update({
      where: { id: product_id },
      data: {
        is_deleted: true,
        last_updated_by: user_id,
        last_updated_at: new Date().toISOString(),
      },
    });
  }
}
