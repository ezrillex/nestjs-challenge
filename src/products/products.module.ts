import { Module } from '@nestjs/common';
import { AuthModule } from '../auth/auth.module';
import { ProductsService } from './products.service';
import { PrismaService } from '../prisma/prisma.service';
import { ProductsResolver } from './products.resolver';
import { ProductVariationResolver } from './product_variation/product_variation.resolver';

@Module({
  imports: [AuthModule],
  providers: [
    ProductsService,
    PrismaService,
    ProductsResolver,
    ProductVariationResolver,
  ],
  exports: [ProductsService],
})
export class ProductsModule {}
