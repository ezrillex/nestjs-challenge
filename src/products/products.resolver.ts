import { Resolver, Query, Mutation, Args, Context } from '@nestjs/graphql';
import { roles } from '@prisma/client';
import { RequiresRole } from '../common/decorators/requires-role.decorator';
import { CreateProductInput } from './inputs/createProduct.input';
import { ProductsService } from './products.service';
import { Products } from './products.model';
import { GetProductsInput } from './inputs/get-products.input';
import { UpdateProductInput } from './inputs/update-product.input';
import { UpdateProductVariationInput } from './product_variation/update-product-variation.input';
import { ParseUUIDPipe } from '@nestjs/common';
import { PublicPrivate } from '../common/decorators/public_and_private.decorator';
import { CreateProductVariationInput } from './product_variation/create_product_variation.input';

@Resolver()
export class ProductsResolver {
  constructor(private productsService: ProductsService) {}

  @PublicPrivate()
  @Query(() => [Products], { nullable: true })
  async getProducts(
    @Args('GetProductsInput') getProductsInput: GetProductsInput,
    @Context('req') request: Request,
  ) {
    // todo on auth guard we set a value to know what mode we are in...
    let role: roles = roles.public;
    if (request['user'] && request['user'].role) {
      role = request['user'].role;
    }
    return this.productsService.GetProducts(role, getProductsInput);
  }

  @PublicPrivate()
  @Query(() => Products, { nullable: true })
  async getProductById(
    @Args('id', { type: () => String }, ParseUUIDPipe) id: string,
    @Context('req') request: Request,
  ) {
    let role: roles = roles.public;
    if (request['user']) {
      role = request['user'].role;
    }
    return this.productsService.GetProductById(role, id);
  }

  @RequiresRole(roles.manager)
  @Mutation(() => String, { nullable: true })
  async createProduct(
    @Args('CreateProductInput') createProductInput: CreateProductInput,
    @Context('req') request: Request,
  ) {
    const result = await this.productsService.CreateProduct(
      createProductInput,
      request['user'].id,
    );
    return result.id;
  }

  @RequiresRole(roles.manager)
  @Mutation(() => String, { nullable: true })
  async updateProduct(
    @Args('UpdateProductInput') updateProductInput: UpdateProductInput,
    @Context('req') request: Request,
  ) {
    const result = await this.productsService.UpdateProduct(
      updateProductInput,
      request['user'].id,
    );
    return result.id;
  }

  @RequiresRole(roles.manager)
  @Mutation(() => String, { nullable: true })
  async updateProductVariation(
    @Args('UpdateProductVariationInput')
    updateProductVariationInput: UpdateProductVariationInput,
    @Context('req') request: Request,
  ) {
    const result = await this.productsService.UpdateProductVariation(
      updateProductVariationInput,
      request['user'].id,
    );
    return result.id;
  }

  @RequiresRole(roles.manager)
  @Mutation(() => String, { nullable: true })
  async createProductVariation(
    @Args('CreateProductVariationInput')
    createProductVariationInput: CreateProductVariationInput,
    @Context('req') request: Request,
  ) {
    // todo check if product exists, or let it fail?

    const result = await this.productsService.CreateProductVariation(
      createProductVariationInput,
      request['user'].id,
    );
    return result.id;
  }

  @RequiresRole(roles.manager)
  @Mutation(() => String, { nullable: true })
  async deleteProductVariation(
    @Args('variation_id', { type: () => String }, ParseUUIDPipe)
    variation_id: string,
  ) {
    return this.productsService.DeleteProductVariation(variation_id);
  }

  @RequiresRole(roles.manager)
  @Mutation(() => String, { nullable: true })
  async deleteProduct(
    @Args('product_id', { type: () => String }, ParseUUIDPipe)
    product_id: string,
    @Context('req') request: Request,
  ) {
    // TODO because the db will cascade when we do the actual delete, we need to first delete the images from CDN.

    const result = await this.productsService.DeleteProduct(
      product_id,
      request['user'].id,
    );
    return result.id;
  }
}
