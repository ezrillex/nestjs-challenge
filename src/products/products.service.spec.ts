import { Test, TestingModule } from '@nestjs/testing';
import { PrismaService } from '../prisma/prisma.service';
import { ProductsService } from './products.service';
import { UpdateProductInput } from './inputs/update-product.input';

describe('LikesService', () => {
  let service: ProductsService;
  let prismaService: PrismaService;

  beforeEach(async () => {
    const module: TestingModule = await Test.createTestingModule({
      providers: [PrismaService, ProductsService],
    }).compile();

    service = module.get<ProductsService>(ProductsService);
    prismaService = module.get<PrismaService>(PrismaService);
  });

  it('should be defined', () => {
    expect(service).toBeDefined();
  });

  it('create product should be defined', () => {
    expect(service.CreateProduct).toBeDefined();
  });

  it('update product should be defined', () => {
    expect(service.UpdateProduct).toBeDefined();
  });

  it('get products should be defined', () => {
    expect(service.GetProducts).toBeDefined();
  });

  it('get products by id should be defined', () => {
    expect(service.GetProductById).toBeDefined();
  });

  it('get products variation should be defined', () => {
    expect(service.GetProductVariationById).toBeDefined();
  });

  it('update products variation should be defined', () => {
    expect(service.UpdateProductVariation).toBeDefined();
  });

  it('create product variation should be defined', () => {
    expect(service.CreateProductVariation).toBeDefined();
  });

  it('delete product variation should be defined', () => {
    expect(service.DeleteProductVariation).toBeDefined();
  });

  it('delete product should be defined', () => {
    expect(service.DeleteProduct).toBeDefined();
  });

  describe('create product', () => {
    it('passes data to prisma query correctly', async () => {
      jest.useFakeTimers();
      jest.setSystemTime(new Date(2020, 12, 12, 12, 12));
      const spy = jest
        .spyOn(prismaService.products, 'create')
        .mockResolvedValue(null);

      await service.CreateProduct(
        {
          name: 'test name',
          description: 'test description',
          categories: ['076b5b00-c719-40c3-a8f2-d1a11c17b75c'],
          variations: [
            { title: 'test var 1', stock: 123, price: 321.11 },
            { title: 'test var 2', stock: 321, price: 321.22 },
          ],
        },
        '076b5b00-c719-40c3-a8f2-d1a11c17b75c',
      );

      expect(spy.mock.calls).toMatchSnapshot('create product query');
    });
  });

  describe('update product', () => {
    it('throws if product is not found', async () => {
      await expect(
        service.UpdateProduct(
          { id: '076b5b00-c719-40c3-a8f2-d1a11c17b75c' } as UpdateProductInput,
          '076b5b00-c719-40c3-a8f2-d1a11c17b75c',
        ),
      ).rejects.toThrowErrorMatchingSnapshot('not found error product');
    });

    it('throws if no values to update are provided', async () => {
      jest.spyOn(prismaService.products, 'count').mockResolvedValue(1);

      await expect(
        service.UpdateProduct(
          { id: '076b5b00-c719-40c3-a8f2-d1a11c17b75c' } as UpdateProductInput,
          '076b5b00-c719-40c3-a8f2-d1a11c17b75c',
        ),
      ).rejects.toThrowErrorMatchingSnapshot('one is required to update');
    });

    it('passes values to the update query', async () => {
      jest.useFakeTimers();
      jest.setSystemTime(new Date(2020, 12, 12, 12, 12));
      jest.spyOn(prismaService.products, 'count').mockResolvedValue(1);
      const spy = jest
        .spyOn(prismaService.products, 'update')
        .mockResolvedValue(null);

      await expect(
        service.UpdateProduct(
          {
            id: '076b5b00-c719-40c3-a8f2-d1a11c17b75c',
            name: 'test name',
            is_published: false,
            description: 'test description',
            categories: ['test category id'],
          },
          '076b5b00-c719-40c3-a8f2-d1a11c17b75c',
        ),
      ).resolves.not.toThrow();

      expect(spy.mock.calls).toMatchSnapshot('prisma query products update');
    });
  });
});
