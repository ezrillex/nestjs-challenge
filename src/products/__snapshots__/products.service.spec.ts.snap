// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`LikesService create product passes data to prisma query correctly: create product query 1`] = `
[
  [
    {
      "data": {
        "categories": {
          "connect": [
            {
              "id": "076b5b00-c719-40c3-a8f2-d1a11c17b75c",
            },
          ],
        },
        "created_at": "2021-01-12T18:12:00.000Z",
        "created_by": "076b5b00-c719-40c3-a8f2-d1a11c17b75c",
        "description": "test description",
        "name": "test name",
        "variations": {
          "createMany": {
            "data": [
              {
                "price": 321.11,
                "stock": 123,
                "title": "test var 1",
              },
              {
                "price": 321.22,
                "stock": 321,
                "title": "test var 2",
              },
            ],
          },
        },
      },
    },
  ],
]
`;

exports[`LikesService update product passes values to the update query: prisma query products update 1`] = `
[
  [
    {
      "data": {
        "categories": {
          "set": [
            {
              "id": "test category id",
            },
          ],
        },
        "description": "test description",
        "last_updated_at": "2021-01-12T18:12:00.000Z",
        "last_updated_by": "076b5b00-c719-40c3-a8f2-d1a11c17b75c",
        "name": "test name",
      },
      "where": {
        "id": "076b5b00-c719-40c3-a8f2-d1a11c17b75c",
      },
    },
  ],
]
`;

exports[`LikesService update product throws if no values to update are provided: not found error product 1`] = `"At least one property is required for update!"`;

exports[`LikesService update product throws if no values to update are provided: one is required to update 1`] = `"At least one property is required for update!"`;

exports[`LikesService update product throws if product is not found: not found error product 1`] = `"Product not found."`;
